cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -388,13 +391,46 @@ private void executeFork(final ApplicationSpecification appSpec, WorkflowForkNod
     }
   }
 
+  private void executeCustomAction(WorkflowActionNode actionNode, InstantiatorFactory instantiator,
+                                   ClassLoader classLoader, WorkflowToken token) throws Exception {
+    String customActionName = actionNode.getProgram().getProgramName();
+    RunId customActionRunId = RunIds.generate();
+    long startTimeInSeconds = RunIds.getTime(customActionRunId, TimeUnit.SECONDS);
+    Id.Program customActionId = Id.Program.from(workflowId.getApplication(), ProgramType.CUSTOM_ACTION,
+                                                customActionName);
+
+    store.setWorkflowProgramStart(customActionId, customActionRunId.getId(), workflowId.getId(), runId.getId(),
+                                  actionNode.getNodeId(), startTimeInSeconds, null);
+
+    ProgramRunStatus runStatus = ProgramRunStatus.COMPLETED;
+
+    try {
+      executeAction(actionNode, instantiator, classLoader, token);
+    } catch (Throwable t) {
+      runStatus = ProgramRunStatus.FAILED;
+      if (Iterables.size(Iterables.filter(Throwables.getCausalChain(t), InterruptedException.class)) > 0) {
+        runStatus = ProgramRunStatus.KILLED;
```

Comment:
For custom actions we do not have program runner/listeners. 
I thought of using the `runningThread` so that
```java
catch (Throwable t) {
   if (runningThread == null) {
      // mark custom action as KILLED    
   }
  else {
     // mark custom action as FAILED
  }
}
```
However above code only handles the explicit `stop` of the Workflow. If custom action is running as a part of the FORK branch and if program on another branch is failed, the custom action need to be marked as KILLED. So I used checks for the `InterruptedException`.